iothub/device/src/Transport/Amqp/AmqpUnit.cs

```diff
@@ -293,95 +328,109 @@ public void OnEventsReceived(Message message)
-                if (Logging.IsEnabled) Logging.Associate(this, _methodSendingLink, $"{nameof(_methodSendingLink)}");
+            if (Logging.IsEnabled) Logging.Enter(this, timeout, $"{nameof(EnableMethodsAsync)}");
+            AmqpIoTSession amqpIoTSession = await EnsureSessionAsync(timeout).ConfigureAwait(false);
+            bool gain = await _methodLinkLock.WaitAsync(timeout).ConfigureAwait(false);
+            if (!gain)
+            {
+                throw new TimeoutException();
             }
-            catch (Exception)
+            
+            string correlationIdSuffix = Guid.NewGuid().ToString();
+            try
             {
-                _methodReceivingLink?.Abort();
-                _methodReceivingLink = null;
-
-                _methodSendingLink?.Abort();
-                _methodSendingLink = null;
-
-                throw;
+                await Task.WhenAll(
+                    OpenMethodsReceiverLinkAsync(amqpIoTSession, correlationIdSuffix, timeout),
+                    OpenMethodsSenderLinkAsync(amqpIoTSession, correlationIdSuffix, timeout)
+                ).ConfigureAwait(false);
             }
             finally
             {
+                _methodLinkLock.Release();
                 if (Logging.IsEnabled) Logging.Exit(this, timeout, $"{nameof(EnableMethodsAsync)}");
             }
         }
-
-        private void OnMethodReceived(MethodRequestInternal methodRequestInternal)
+        
+        private async Task OpenMethodsReceiverLinkAsync(AmqpIoTSession amqpIoTSession, string correlationIdSuffix, TimeSpan timeout)
         {
-            if (Logging.IsEnabled) Logging.Enter(this, methodRequestInternal, $"{nameof(OnMethodReceived)}");
-            try
+            if (_methodReceivingLink == null || _methodReceivingLink.IsClosing())
             {
-                _methodHandler?.Invoke(methodRequestInternal);
+                _methodReceivingLink = await amqpIoTSession.OpenMethodsReceiverLinkAsync(_deviceIdentity, correlationIdSuffix, timeout).ConfigureAwait(false);
+                _methodReceivingLink.Closed += (obj, arg) => {
+                    amqpIoTSession.SafeClose();
+                };
+                _methodReceivingLink.RegisterMethodListener(OnMethodReceived);
+                if (Logging.IsEnabled) Logging.Associate(this, _methodReceivingLink, $"{nameof(_methodReceivingLink)}");
             }
-            finally
+        }
+        private async Task OpenMethodsSenderLinkAsync(AmqpIoTSession amqpIoTSession, string correlationIdSuffix, TimeSpan timeout)
+        {
+            if (_methodSendingLink == null || _methodSendingLink.IsClosing())
             {
-                if (Logging.IsEnabled) Logging.Exit(this, methodRequestInternal, $"{nameof(OnMethodReceived)}");
+                _methodSendingLink = await amqpIoTSession.OpenMethodsSenderLinkAsync(_deviceIdentity, correlationIdSuffix, timeout).ConfigureAwait(false);
```

Comment:
Same here and everywhere else where links are opened. The pattern has a race condition.

Reply:
Same as above. Table P0